stacks-signer/src/client/stackerdb.rs

```diff
@@ -0,0 +1,312 @@
+                Ok(StackerDBMessage::Block(block))
+            }
+        }
+    }
+}
+
+impl StackerDBMessage {
+    /// Helper function to determine the slot ID for the provided stacker-db writer id
+    pub fn slot_id(&self, id: u32) -> u32 {
+        let slot_id = match self {
+            StackerDBMessage::Packet(packet) => match packet.msg {
+                Message::DkgBegin(_) => DKG_BEGIN_SLOT_ID,
+                Message::DkgPrivateBegin(_) => DKG_PRIVATE_BEGIN_SLOT_ID,
+                Message::DkgEnd(_) => DKG_END_SLOT_ID,
+                Message::DkgPublicShares(_) => DKG_PUBLIC_SHARES_SLOT_ID,
+                Message::DkgPrivateShares(_) => DKG_PRIVATE_SHARES_SLOT_ID,
+                Message::NonceRequest(_) => NONCE_REQUEST_SLOT_ID,
+                Message::NonceResponse(_) => NONCE_RESPONSE_SLOT_ID,
+                Message::SignatureShareRequest(_) => SIGNATURE_SHARE_REQUEST_SLOT_ID,
+                Message::SignatureShareResponse(_) => SIGNATURE_SHARE_RESPONSE_SLOT_ID,
+            },
+            Self::Block(_block) => BLOCK_SLOT_ID,
+        };
+        SIGNER_SLOTS_PER_USER * id + slot_id
+    }
+}
+/// The StackerDB client for communicating with both .signers and .miners contracts
+pub struct StackerDB {
+    /// The stacker-db session for the signer StackerDB
+    signers_stackerdb_session: StackerDBSession,
+    /// The stacker-db session for the .miners StackerDB
+    miners_stackerdb_session: StackerDBSession,
+    /// The private key used in all stacks node communications
+    stacks_private_key: StacksPrivateKey,
+    /// A map of a slot ID to last chunk version
+    slot_versions: HashMap<u32, u32>,
+}
+
+impl From<&Config> for StackerDB {
+    fn from(config: &Config) -> Self {
+        Self {
+            signers_stackerdb_session: StackerDBSession::new(
+                config.node_host,
+                config.signers_stackerdb_contract_id.clone(),
+            ),
+            miners_stackerdb_session: StackerDBSession::new(
+                config.node_host,
+                config.miners_stackerdb_contract_id.clone(),
+            ),
+            stacks_private_key: config.stacks_private_key,
+            slot_versions: HashMap::new(),
+        }
+    }
+}
+
+impl StackerDB {
+    /// Sends messages to the stacker-db with an exponential backoff retry
+    pub fn send_message_with_retry(
+        &mut self,
+        id: u32,
+        message: StackerDBMessage,
+    ) -> Result<StackerDBChunkAckData, ClientError> {
+        let message_bytes = message.serialize_to_vec();
+        let slot_id = message.slot_id(id);
+
+        loop {
+            let slot_version = *self.slot_versions.entry(slot_id).or_insert(0) + 1;
+            let mut chunk = StackerDBChunkData::new(slot_id, slot_version, message_bytes.clone());
+            chunk.sign(&self.stacks_private_key)?;
+            debug!("Sending a chunk to stackerdb!\n{:?}", chunk.clone());
```

Comment:
Nit: You probably don't need the clone here; you could just pass `&chunk`.

Reply:
Oh yes definitely here :)